### PR TITLE
feat(handle_state): token-borne --list-reserved mode; hs_finalize_token (closes #143)

### LINKED ISSUE #143: [Feature] handle_state: carry --list-reserved mode in the state token
## Summary

Carry the `--list-reserved` operating mode inside the state token value (an HS2 sentinel record) instead of declaring a `list_reserved` local in the entry-point frame, and make the generated eval-code capture the *caller's* frame so the reported collision surface is end-to-end accurate.

**Supersedes #139** — that issue asked for the collision surface to be returned as the `list_reserved` *value*; this design returns it as the *token* value and removes the `list_reserved` name from the pattern entirely.

## Motivation

Three defects in the current eval-code `--list-reserved` design:

1. **Namespace pollution.** `--list-reserved` is a global operating mode, yet it is signalled by declaring a plain-named local (`list_reserved`) in the entry-point frame. The `-S` pattern's whole point is to minimise the names an entry point injects; the mode marker itself violates that.
2. **Incomplete capture.** `hs_extract_token` emits a *frozen* name list computed from its own subshell frame. Locals the API entry point declared *before* the eval — precisely the pattern violations `--list-reserved` exists to expose — are invisible, because `local -p` only enumerates the current function's frame and only the emitted code runs in that frame. The report can therefore under-state the real collision surface.
3. **Unreadable detection idiom.** Entry points test the mode with `local -p list_reserved >/dev/null 2>&1` (or `${list_reserved@A}`), which reads as plumbing, not intent.

### Token-borne mode marker

- `hs_extract_token`'s eval-code `--list-reserved` form emits code that sets the **token local** to an HS2 sentinel record, e.g. `HS2:mode=list_reserved:<encoded names>` (exact encoding to be fixed at planning; HS2 format discipline guarantees no legitimate token can equal it, and every other consumer — `hs_read_persisted_state` etc. — rejects it with a discriminable error instead of proceeding).
- No `local list_reserved` is declared anywhere. The only name an entry point ever declares is its token local.
- `hs_write_token`'s **normal path** reads `${!2}` through dynamic scope, detects the sentinel, and switches to merge-and-report emission. The entry point no longer needs a mode branch in its argument handling: the pattern becomes uniformly *extract → body-unless-listing → write*.

### Convenience predicate

New public function (name to be fixed at planning), e.g.:

```bash
hs_token_lists_reserved <token_local_name>   # returns 0 iff the token holds the mode sentinel
```

Canonical entry-point skeleton becomes:

```bash
mod_func() {
    eval "$(hs_extract_token mod_func __mod_tok "$@")" || return $?
    hs_token_lists_reserved __mod_tok || { _mod_func "$@" || return $?; }
    eval "$(hs_write_token mod_func __mod_tok "$@")" || return $?
}
```

### End-to-end frame capture in the generated code

- The emitted eval-code performs `local -p` **in the entry-point frame** (the only place the caller's locals are enumerable) and merges that capture with the frozen subshell-side names before encoding into the token.
- Ordering: the token local is declared first (it is legitimately reserved), then the capture runs, then the sentinel value is assigned — so no receiver variable ever appears in its own capture and nothing can be masked.
- `hs_write_token`'s emitted report performs a **fresh** capture at write time, so locals declared between the two evals (also pattern violations) are reported too. Generated code must deduplicate the merged set and exclude only the machinery's own transient names.
- Proven mechanics: the `local receiver="$(local -p)"` self-exclusion trick is already used inside the library.

## Acceptance Criteria

- [ ] No `list_reserved` local is declared by any generated code; the mode marker lives in the token value as an HS2 sentinel record
- [ ] Real tokens can never equal the sentinel; all non-token-utility consumers reject it with a discriminable error code
- [ ] `hs_token_lists_reserved` (final name TBD) added, documented, and used in the canonical skeletons
- [ ] `hs_extract_token` eval-code output captures the entry-point frame at eval time; a stray `local leaked=1` declared before the eval appears in the report (new regression test)
- [ ] `hs_write_token` report captures the frame at write time; a stray local declared between the evals appears in the report (new regression test)
- [ ] Exact-set-equality test updated: report == `hs_persist_state --list-reserved` ∪ {token local} with no sentinel special-casing
- [ ] Reserved-name rejection, entry-point pattern docs (RST), skill, and templates updated
- [ ] `--list-reserved` re-entry safety preserved (no recursion through the guard query; see the PR #140 fork-bomb history)

## Alternatives Considered

- **Keep the `list_reserved` sentinel and only fix the capture** (#139 direction): retains the namespace pollution and the unreadable detection idiom; rejected.
- **Environment variable or global as mode marker**: leaks across frames and subshells, worse than a local; rejected.
- **Separate reporting function instead of a mode**: breaks the "one forwarded `\"$@\"` drives everything" convention of the entry-point pattern; rejected.

## Additional Context

- Builds on the entry-point pattern landed by #136 / PR #140; must branch after #140 merges.
- Breaking change to the documented entry-point pattern (sentinel detection idiom disappears); existing in-repo entry points and tests migrate in the same change.
- Origin: design discussion of 2026-07-11 (skill/doc articulation and collision-surface review).

### PR DESCRIPTION
**Pull request by lepeuvedic-natixar**
## Description

Carries the `--list-reserved` operating mode inside the state **token** (an HS2 mode-token record) instead of a `list_reserved` local in the entry-point frame, and makes the generated eval-code capture the caller's frame end-to-end so the reported collision surface is accurate. Closes #143.

**Breaking change** to the documented entry-point pattern: the `hs_write_token` name and the `local -p list_reserved` detection idiom are gone.

## Type of Change
- Breaking change / feature (`config/handle_state.sh`)

## What changed
- **Mode token**: `hs_extract_token --list-reserved` mints `HS2:mode=list-reserved:<reserved_names payload>` (built by `hs_persist_state`, checksum field swapped for the marker). The token local is declared late, so it is absent from its own capture.
- **`hs_write_token` → `hs_finalize_token`** — always the last line of an entry point, token-driven:
  - mode token → emit the collision report (`reserved_names` + a **fresh** entry-point-frame capture, adding the token local unless the marker ends in `-ro`);
  - normal token with `-S` → write back; without `-S` → silent read-only no-op.
- **`hs_is_list_reserved_mode -S <token>`** — body-skip guard.
- **`hs_read_only <API> <token> "$@"`** — optional read-only marker; appends `-ro` to any `mode=` marker (future-proof), else strips `-S` from the forwarded args.
- **`HS_ERR_LIST_RESERVED_TOKEN=13`** — `_hs_hs2_parse` rejects any `mode=` token discriminably (structural, printed), fencing off `hs_persist_state` / `hs_destroy_state` / `hs_read_persisted_state`.
- Single canonical entry-point skeleton; RST + skill (trimmed to a tool list) updated.
- SCM Step 1 doc: posting the assessment as an issue comment is now a mandatory deliverable.

## Testing
- `test/test-hs_persist_state.bats`: **131 tests pass locally** (converted the former `hs_write_token`/`list_reserved` tests; added mode-token minting, predicate, dynamic-scope write-back, end-to-end frame-capture, discriminable-rejection, read-only, and edge-case tests).
- `sphinx-build -b html` builds clean.

## Related Issues
Closes #143. Builds on #140 (entry-point token utilities).

## Acceptance criteria
- [x] No `list_reserved` local declared by generated code; marker lives in the token value
- [x] Real tokens can never equal the marker; consumers reject with a discriminable code
- [x] `hs_is_list_reserved_mode` added, documented, used in the skeleton
- [x] eval-time capture (stray local before the eval appears in the report) — regression test
- [x] write-time capture (stray local between the evals appears) — regression test
- [x] Exact-set-equality test updated (report == `hs_persist_state --list-reserved` ∪ {token local})
- [x] Reserved-name rejection, entry-point docs (RST), skill, and templates updated
- [x] `--list-reserved` re-entry safety preserved

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: config/handle_state.sh [hunk 1]
```diff
@@ -525,39 +530,126 @@ hs_write_token() {
...
+            printf '\nreturn 0\n'
             return 0
         fi
-        # Eval-code --list-reserved form: compute own surface, merge with
-        # list_reserved from the entry-point frame (set by hs_extract_token in
-        # read-write patterns), add source-local, emit printf + return 0.
+    fi
+    # Normal token: write back only when the caller passed -S; a missing -S means
+    # the entry point is read-only w.r.t. external state (silent no-op).
+    local __hs_ft_has_s=0 __hs_ft_arg
```
Comment (lepeuvedic) [inline config/handle_state.sh:598]: The special purpose options parser here is a design error.
If _hs_resolve_state_input requires -S , make it more flexible with an option flag.

This is code duplication and a poor implementation.

File: config/handle_state.sh [hunk 2]
```diff
@@ -525,39 +530,126 @@ hs_write_token() {
...
-        done <<< "$hs_wt_names"
-        if [[ -v list_reserved ]]; then
-            while IFS= read -r _hs_name; do
-                [[ -n "$_hs_name" ]] && _hs_merged["$_hs_name"]=1
-            done <<< "$list_reserved"
+        _hs_print_reserved_names "$lp_snapshot"
+        return 0
+    fi
+    _hs_resolve_state_inputs hs_is_list_reserved_mode S: "$@" || return $?
+    local __hs_ilrm_tok="${!__hs_processed[state]}"
```
Comment (lepeuvedic) [inline config/handle_state.sh:634]: Stronger guarantees should be provided by _hs_resolve_state_inputs regarding __hs_processed[state]. It would avoid repeated check code.

**Review by lepeuvedic — APPROVED** [review]

File: config/handle_state.sh (hunk 1 above)
Comment (lepeuvedic-natixar) [reply config/handle_state.sh:598]: Suivi par #152 — l'analyseur ad hoc est supprimé au profit d'un drapeau rendant `-S` facultatif dans `_hs_resolve_state_inputs`. L'issue documente en plus la conséquence mesurée : les formes groupées (`-bS <var>`) échappent au scan littéral et la réécriture du jeton est perdue sans diagnostic. Hors périmètre de #143 ; ce fil peut être résolu.

File: config/handle_state.sh (hunk 2 above)
Comment (lepeuvedic-natixar) [reply config/handle_state.sh:634]: Suivi par #153 — renforcement du contrat de `_hs_resolve_state_inputs` sur `__hs_processed[state]`, avec suppression du bloc de contrôle dupliqué dans `_hs_rr_explicit_stmts` / `_hs_rr_implicit_snippet` et des trois lectures indirectes non vérifiées. Hors périmètre de #143 ; ce fil peut être résolu.
